gquant/dataframe_flow/node.py

```diff
@@ -1,64 +1,159 @@
 import abc
-import numpy as np
 import os
 import cudf
-import pandas as pd
-import dask_cudf
-import dask
 
-OUTPUT_ID = 'f291b900-bd19-11e9-aca3-a81e84f29b0f_uni_output'
+from .task import Task
+from .taskSpecSchema import TaskSpecSchema
+from .portsSpecSchema import PortsSpecSchema
 
-__all__ = ['Node', 'OUTPUT_ID']
+from ._node import _Node
 
 
-class Node(object):
-    __metaclass__ = abc.ABCMeta
+__all__ = ['Node']
+
+
+class _PortsMixin(object):
+    '''Mixed class must have (doesn't have to implement i.e. relies on
+    NotImplementedError) "ports_setup" method otherwise raises AttributeError.
```

Comment:
Why not move the ports_setup into this class?

Reply:
Here I'm trying to have a clear separation between public interface methods i.e. API of `Node` class for end-user and implementation details. In the `Node` class the methods that an end-user will have to override are:
```
ports_setup - required if using ports API
columns_setup - required for dataframe columns type matching and validation
process - actual processing logic
```
Additionally I added methods in the `Node` class:
```
load_cache
save_cache
```
The cache methods are in case the end-user wants a customized caching scheme. I need to add docstrings to these methods.

The other methods are:
```
using_ports
get_input_ports
get_output_ports
```
Which delegate to their corresponding methods in the `_PortsMixin` class. Including these last three methods in public API is probably a mistake and I should not include them explicitly to the `Node` class (I'll remove them from `Node`) . I can just leave them in the mixin class and from an end-users perspective they are somewhat hidden as they are an implementation detail. The public API is `ports_setup`.

Anything that an end-user has to work with directly should be in the `Node` class, and other stuff can be in mixin classes. I want the generated docs and help on the `Node` class to only show what the end-user implementing a custom node plugin should be concerned with and nothing more. In the `05b_customize_nodes_with_ports.ipynb` only three methods are shown for defining a node: `ports_setup`, `columns_setup`, `process`; therefore these methods are part of the public `Node` API.

Reply:
Yes. It is much cleaner this way.

Is `_PortsMixin` necessary? Is it introduced to be different from the old no port graph?

Reply:
The `_PortsMixin` is for ISP. I don't want to clutter up the `Node` class with methods used for implementation details rather than interface API. I want the `Node` class to be an API abstract base class as far the end-user/client is concerned. Therefore except for `ports_setup` I want all the other helper methods for checking and accessing ports to be hidden.
I removed the methods: `using_ports`, `get_input_ports`, `get_output_ports` from `Node`. Now the `help(Node)` looks cleaner. The client should just see the docstrings and methods relevant to implementing their own `Node`.